### PR TITLE
Stripe: Sprout Initial Stripe Utility

### PR DESCRIPTION
**Pull request by zspencer**
https://github.com/zinc-collective/convene/issues/831 https://github.com/zinc-collective/convene/issues/1002

This is mostly because I want to be able to move forward with connecting Vendor and Distributor accounts in the Marketplace, and to get started fiddling with the Stripe Connect bits.

This also hopefully helps set the stage to getting rid of `Utility` classes, and instead have those inherit fro `UtilityHookup` (kind of like how new `Furniture` inherits from `FurniturePlacement`

To be honest, I should probably either do a rename pass to more clearly indicate the old-way vs now-and-forward way of defining furniture, or do a final cleanup pass on both of those before starting this, but this was the nearest-mess-to-muck-with ( not clean... just.. uhhh... muck with)

Anyway!

### CODE REVIEW

File: app/policies/utility_hookup_policy.rb [hunk 1]
```diff
@@ -41,8 +41,8 @@ def destroy?
 
   def permitted_attributes(params)
     utility_permitted_attributes =
-      policy(Utilities.new_from_slug(params[:utility_slug]))
-        .permitted_attributes(params[:utility_attributes])
+      policy!(Utilities.new_from_slug(params[:utility_slug]&.to_sym))
```
Comment (zspencer) [inline app/policies/utility_hookup_policy.rb:44]: Could use the `Utilities.fetch` here!

File: app/models/utility_hookup.rb [hunk 2]
```diff
@@ -46,4 +46,21 @@ def utility_attributes=(attributes)
...
+    "#{self.class.name.demodulize.underscore}/form"
+  end
+
+  def self.from_utility_hookup(utility_hookup)
+    utility_hookup.becomes(self)
+  end
+
+  def polymorph
+    x = Utilities::REGISTRY.fetch(utility_slug&.to_sym, NullUtility)
+    x.ancestors.include?(UtilityHookup) ? becomes(x) : self
```
Comment (zspencer) [inline app/models/utility_hookup.rb:64]: This name is lol ok no. Also could use `Utilities.fetch`!

File: app/models/utility_hookup.rb [hunk 3]
```diff
@@ -24,8 +25,7 @@ def name
   attribute :status, :string, default: "unavailable"
   validates :status, presence: true, inclusion: {in: %w[ready unavailable]}
 
-  attribute :old_configuration, :json, default: -> { {} }
-  validates_associated :utility
+  # validates_associated :utility
```
Comment (zspencer) [inline app/models/utility_hookup.rb:28]: I pretty sure I want to bring this back but it makes a form go sad somewhere so I need to be less stoned to figure it out.

**Comment by zspencer** [comment]
CI Failed, back-to-draft with ye.

Thoughts on the trajectory v much welcome.

File: app/utilities/utility.rb [hunk 4]
```diff
@@ -25,4 +25,8 @@ def form_template
   def display_name
     model_name.human.titleize
   end
+
+  def self.from_utility_hookup(utility_hookup)
+    new(utility_hookup: utility_hookup)
+  end
```
Comment (zspencer) [inline app/utilities/utility.rb:31]: Delete this

File: app/utilities/stripe_utility.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+class StripeUtility < UtilityHookup
```
Comment (zspencer) [inline app/utilities/stripe_utility.rb:1]: Probably should just inherit from utility :/

**Comment by zspencer** [comment]
This is an intermediary step to at least get the Stripe API Token settable; but I pulled back from refactoring the Utilities structure.

I'm going to try and finish unfucking the Furniture structure(s) before I start unfucking Utility.

**Comment by anaulin** [comment]
@zspencer do you still want to merge this Utility, or are we now preferring the approach in https://github.com/zinc-collective/convene/pull/1009? Happy to review this if you still want a review.

**Comment by zspencer** [comment]
I'm not entirely sure, tbh. I think we have open questions about whether the Stripe API Key wants to live in the Marketplace, or at the Space level. Storing it at the Marketplace for now is probably the straightest path towards a working feature; but having the Utility available/merged in could be useful for when a Space has many Marketplaces.

File: app/models/utility_hookup.rb [hunk 6]
```diff
@@ -9,9 +9,10 @@ class UtilityHookup < ApplicationRecord
   # has multiple {UtilityHookup}s.
   # @return [String]
   attribute :name, :string
+  validates :name, presence: :true, uniqueness: { scope: :space_id }
```
Comment (anaulin) [inline app/models/utility_hookup.rb:12]: This is ok for now, but I can imagine wanting to have more than one Stripe account associated with a space in the future. But maybe the hookup itself will support multiple API keys. We'll cross that bridge when we get to it.

**Review by anaulin — APPROVED** [review]
💰 🔑 ✅

File: app/models/utility_hookup.rb (hunk 6 above)
Comment (zspencer) [reply app/models/utility_hookup.rb:12]: This is validating the name given to the utility to disambiguate it against different utilities on the same space. I.e. `Stripe Account - Jim's Beans` vs `Stripe Account - Jane's Jams`

Comment (anaulin) [reply app/models/utility_hookup.rb:12]: Oh 🤦🏼‍♀️ Bad code reviewer, no donut for me.

**Pull request merged**